client/core/trade.go

```diff
@@ -1612,90 +1630,107 @@ func (c *Core) redeemMatchGroup(t *trackedTrade, matches []*matchTracker, errs *
 		)
 	}
 
-	// Send the redemption information to the DEX.
+	// Save redemption details and send the redeem message to the DEX.
+	// Saving the redemption details now makes it possible to resend the
+	// `redeem` reqeust at a later time if sending it now fails.
 	for i, match := range matches {
-		err := c.finalizeRedeemAction(t, match, coinIDs[i])
-		if err != nil {
-			errs.addErr(err)
+		proof := &match.MetaData.Proof
+		coinID := []byte(coinIDs[i])
+		if match.Match.Side == order.Taker {
+			match.SetStatus(order.MatchComplete) // could this cause the match to be retired before the `redeem` request succeeds?
+			proof.TakerRedeem = coinID
+		} else {
+			match.SetStatus(order.MakerRedeemed)
+			proof.MakerRedeem = coinID
+		}
+		if err := t.db.UpdateMatch(&match.MetaMatch); err != nil {
+			errs.add("error storing swap details in database for match %s, coin %s: %v",
+				match.id, coinIDString(t.wallets.fromAsset.ID, coinID), err)
 		}
+
+		c.sendRedeemAsync(t, match, coinIDs[i], proof.Secret)
 	}
 }
 
-// finalizeRedeemAction sends a `redeem` request for the specified match,
-// waits for and validates the server's acknowledgement, then saves the
-// redeem details to db.
-// The redeem details are always saved even if sending the `redeem` request
-// errors or a valid ack is not received from the server. This makes it
-// possible to resend the `redeem` request at a later time.
+// sendInitAsync starts a goroutine to send an `redeem` request for the specified
+// match and save the server's ack sig to db. Sends a notification if an error
+// occurs while sending the request or validating the server's response.
 //
 // This method modifies match fields and MUST be called with the trackedTrade
 // mutex lock held for writes.
-func (c *Core) finalizeRedeemAction(t *trackedTrade, match *matchTracker, coinID []byte) error {
-	_, _, proof, auth := match.parts()
-	if len(auth.RedeemSig) != 0 {
-		return fmt.Errorf("'redeem' already sent for match %v", match.id)
+func (c *Core) sendRedeemAsync(t *trackedTrade, match *matchTracker, coinID, secret []byte) {
+	if match.sendingRedeemAsync || match.MetaData.Proof.IsRevoked() {
+		return // no need sending `redeem` request for a revoked match
```

Comment:
For both new `send{Init,Redeem}Async` functions, it would be straightforward to _not_ require the tracked trade mutex lock held, and instead convert the `matchTracker.sending{Init,Redeem}Async` bool fields to uint32 atomics.  The reasoning is to reduce lock contention, which is especially important since the `trackedTrade`'s mutex is for all of its `matchTrackers`, which is unnecessarily too broad in most cases and definitely in this one.  Although, I see that both callers already have the `trackedTrade` locked anyway, so maybe we don't make this change unless we want to start making these functions available to callers who may not want to or need to lock the mutex.

So if we did the above:

```go
// occurs while sending the request or validating the server's response.
func (c *Core) sendRedeemAsync(t *trackedTrade, match *matchTracker, coinID, secret []byte) {
	if !atomic.CompareAndSwapUint32(&match.sendingRedeemAsync, 0, 1) {
		return
	}

	c.log.Infof(...)

	// Send the redeem request asynchronously.
	go func() {
		var err error
		defer func() {
			atomic.StoreUint32(&match.sendingRedeemAsync, 0)
			if err != nil {
				corder := t.coreOrder()
```

(Similar for `sendInitAsync`)

decls:

```go
type matchTracker struct {
	// sendingInitAsync indicates if this match's init request is being sent to
	// the server and awaiting a response. No attempts will be made to send
	// another init request for this match while one is already active.
	sendingInitAsync uint32 // atomic
	// sendingRedeemAsync indicates if this match's redeem request is being sent
	// to the server and awaiting a response. No attempts will be made to send
	// another redeem request for this match while one is already active.
	sendingRedeemAsync uint32 // atomic

	db.MetaMatch
	id order.MatchID
```

Reply:
The `match.MetaData.Proof.IsRevoked()` requires the lock to be held currently too.